### PR TITLE
registry: integrate layer info cache with registry and storage

### LINKED ISSUE #262: Add caching for Layer.Fetch to reduce backend round trips
While fast filesystems and local backend drivers should support fast round trips, it's possible to have a slow backend that make HEAD requests very slow. Reducing the round trips is possible but not desirable due to the current VFS-style backend model. This data is immutable and thus has good caching properties.

### PR DESCRIPTION
**Pull request by stevvooe**
This PR integrates a layer info cache with the registry webapp and storage backend. The main benefit is to cache immutable layer meta data, reducing backend roundtrips. The cache can be configured to use either redis or an inmemory cache.

This provides massive performance benefits for HEAD http checks on layer blobs and manifest verification.

For this, we've defined the interface for layer info caches. Layer info caches speed up access to layer meta data accessed in storage driver backends. The two main operations are tests for repository membership and resolving path and size information for backend blobs.

Two implementations are available. The main implementation leverages redis to store layer info. An alternative implementation simply caches layer info in maps, which should speed up resolution for less sophisticated implementations.

Redis has been integrated with the web application for use with various services. The configuration exposes connection details, timeouts and pool parameters. Documentation has been updated accordingly.

A few convenience methods have been added to the context package to get loggers with certain fields, exposing some missing functionality from logrus.

Redigo has been added as a dependency to Godeps.

Closes #262.

### CODE REVIEW

**Comment by stevvooe** [comment]
Supersedes #304.

File: registry/storage/cache/cache.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+	// Meta provides the location of the layer on the backend and its size. Membership of a
+	// repository should be tested before using the result, if required.
+	Meta(ctx context.Context, dgst digest.Digest) (LayerMeta, error)
+
+	// SetMeta sets the meta data for the given layer.
+	SetMeta(ctx context.Context, dgst digest.Digest, meta LayerMeta) error
+}
+
+// base implements common checks between cache implementations. Note that these are not
```
Comment (RichardScothern) [inline registry/storage/cache/cache.go]: not..?

**Comment by stevvooe** [comment]
Added ability to disable caching.

Also did an analysis for memory usage of the inmemory cache. Here we have the total memory usage in the layer cache by number of unique layers, with a repository for each layer:

|  Layers  |   In Memory  | 
|----------|------|
|1,000     |836KB |
|10,000    |8.4MB |
|100,000   |82MB  |
|500,000   |418MB |
|1,000,000 |977MB |

This probably isn't quite an upper bound but is close. There may be a better test that does more layers per repository, but this seems pretty reasonable.

Obviously, redis will be a lot more efficient.

**Comment by dmp42** [comment]
LGTM

**Comment by RichardScothern** [comment]
LGTM

**Pull request merged**